docs/modules/otaupgrade.md

```diff
@@ -0,0 +1,176 @@
+
+If the boot loader is built with rollback support, an extra step is needed
+after the new application has booted (and been tested to be "good", by
+whatever metric(s) the user chooses):
+
+- `otaupgrade.accept()` to mark this image as valid, and allow it to be
+  booted into again.
+
+# otaupgrade.commence()
+
+Wipes the spare application partition and prepares to receive the new
+application firmware.
+
+#### Syntax
+`otaupgrade.commence()`
+
+#### Parameters
+None.
+
+#### Returns
+`nil`
+
+A Lua error may be raised if the OTA upgrade cannot be commenced for some
+reason (such as due to incorrect partition setup).
+
+
+# otaupgrade.write(data)
+
+Write a chunk of application firmware data to the correct partition and
+location. Data must be streamed sequentially, the IDF does not support
+out-of-order data as would be the case from e.g. bittorrent.
+
+#### Syntax
+`otaupgrade.write(data)`
+
+#### Parameters
+- `data` a string of binary data
+
+#### Returns
+`nil`
+
+A Lua error may be raised if the data can not be written, e.g. due to the
+data not being a valid OTA image (the IDF performs some checks in this
+regard).
+
+
+# otaupgrade.complete(reboot)
+
+Finalises the upgrade, and optionally reboots into the new application
+firmware right away.
+
+#### Syntax
+`otaupgrade.complete(reboot)`
+
+#### Parameters
+- `reboot` 1 to reboot into the new firmware immediately, nil to keep running
+
+#### Returns
+`nil`
+
+A Lua error may be raised if the image does not pass validation, or no data
+has been written to the image at all.
+
+
+# otaupgrade.accept()
+
+When the installed boot loader is built with rollback support, a new
+application image is by default only booted once. During this "test run"
+it can perform whatever checks is appropriate (like testing whether it
+can still reach the update server), and if satisfied can mark itself
+as valid. Without being marked valid, upon the next reboot the system
+would "roll back" to the previous version instead.
+
+#### Syntax
+`otaupgrade.accept()`
+
+#### Parameters
+None.
+
+#### Returns
+`nil`
+
+
+# otaupgrade.rollback()
+
+If the installed boot loader is built with rollback support, a new
+firmware may decide that it is not performing as expected, and request
+an explicit rollback to the previous version. If the call to this
+function succeeds, the system will reboot without returning from the
+call.
+
+Note that it is also possible to roll back to a previous firmware
+version even after the new version has called `otaupgrade.accept()`.
+
+#### Syntax
+`otaupgrade.rollback()`
+
+#### Parameters
+None.
+
+#### Returns
+Never. Either the system is rebooted, or a Lua error is raised (e.g. due
+to there being no other firmware to roll back to).
+
+
+# otaupgrade.info()
+
+The boot info and application state and version info can be queried with
+this function. Typically it will be used to check the version of the
+running application.
+
```

Comment:
I was unfamiliar with the IDF's OTA support and so had a number of questions while reading the doc; perhaps it would make sense to front discussion of `otaupgrade.info()` and/or an example of what a typical OTA-ing application would do at boot?

Reply:
As in, move some of the `info()` notes up into the main discussion, and expand on that section?

Reply:
I think I like what you've done in the latest commit. :)